File: nodejs/de_module.js
const dgram = require('dgram');
const { EventEmitter } = require('events');
const AsyncLock = require('async-lock');

const { ANDRUAV_PROTOCOL_TARGET_ID, ANDRUAV_PROTOCOL_MESSAGE_TYPE, ANDRUAV_PROTOCOL_MESSAGE_CMD, INTERMODULE_ROUTING_TYPE, CMD_COMM_SYSTEM, CMD_COMM_GROUP, CMD_COMM_INDIVIDUAL, CMD_TYPE_INTERMODULE, JSON_INTERMODULE_MODULE_KEY, JSON_INTERMODULE_MODULE_ID, JSON_INTERMODULE_MODULE_CLASS, JSON_INTERMODULE_MODULE_MESSAGES_LIST, JSON_INTERMODULE_MODULE_FEATURES, JSON_INTERMODULE_HARDWARE_ID, JSON_INTERMODULE_HARDWARE_TYPE, JSON_INTERMODULE_VERSION, JSON_INTERMODULE_RESEND, JSON_INTERMODULE_TIMESTAMP_INSTANCE, TYPE_AndruavModule_ID, TYPE_AndruavMessage_DUMMY, SPECIAL_NAME_SYS_NAME } = require('./messages.js'); // Adjust path as necessary
const CUDPClient = require('./udpClient'); // Adjust path as necessary

const MODULE_FEATURE_RECEIVING_TELEMETRY = "R";
const MODULE_FEATURE_SENDING_TELEMETRY = "T";
const MODULE_FEATURE_CAPTURE_IMAGE = "C";
const MODULE_FEATURE_CAPTURE_VIDEO = "V";

const MODULE_CLASS_COMM = "comm";
const MODULE_CLASS_FCB = "fcb";
const MODULE_CLASS_VIDEO = "camera";
const MODULE_CLASS_P2P = "p2p";
const MODULE_CLASS_GENERIC = "gen";

const HARDWARE_TYPE_UNDEFINED = 0;
const HARDWARE_TYPE_CPU = 1;

class CModule extends EventEmitter {
    constructor() {
        super();
        if (CModule._instance) {
            return CModule._instance;
        }
        CModule._instance = this;

        this.m_module_class = "";
        this.m_module_id = "";
        this.m_module_key = "";
        this.m_module_version = "";
        this.m_message_filter = {};
        this.cUDPClient = null;
        this.m_party_id = "";
        this.m_group_id = "";
        this.m_OnReceive = null;
        this.m_stdinValues = {};
        this.m_FirstReceived = false;
        this.m_module_features = [];
        this.m_hardware_serial = "";
        this.m_hardware_serial_type = "";
        this.m_instance_time_stamp = Date.now();
        this.m_lock = new AsyncLock(); // Use async-lock for reentrant locking
        return CModule._instance;
    }

    init(target_ip, broadcasts_port, host, listening_port, chunk_size) {
        this.cUDPClient = new CUDPClient();
        this.cUDPClient.init(target_ip, broadcasts_port, host, listening_port, chunk_size, this.onReceive.bind(this));
        this.createJSONID(true);
        this.cUDPClient.start();
        return true;
    }

    uninit() {
        this.cUDPClient.stop();
        return true;
    }

    defineModule(module_class, module_id, module_key, module_version, message_filter) {
        this.m_module_class = module_class;
        this.m_module_id = module_id;
        this.m_module_key = module_key;
        this.m_module_version = module_version;
        this.m_message_filter = message_filter;
    }

    addModuleFeatures(feature) {
        this.m_module_features.push(feature);
    }

    setHardware(hardware_serial, hardware_serial_type) {
        this.m_hardware_serial = hardware_serial;
        this.m_hardware_serial_type = hardware_serial_type;
    }

    sendMSG(msg, length) {
        this.cUDPClient.sendMSG(msg, length);
    }

    sendSysMsg(jmsg, andruav_message_id) {
        const full_message = {
            [ANDRUAV_PROTOCOL_TARGET_ID]: SPECIAL_NAME_SYS_NAME,
            [INTERMODULE_ROUTING_TYPE]: CMD_COMM_SYSTEM,
            [ANDRUAV_PROTOCOL_MESSAGE_TYPE]: andruav_message_id,
            [ANDRUAV_PROTOCOL_MESSAGE_CMD]: jmsg
        };
        const msg = JSON.stringify(full_message);
        this.sendMSG(Buffer.from(msg), msg.length);
    }

    sendJMSG(targetPartyID, jmsg, andruav_message_id, internal_message) {
        this.m_lock.acquire('lock', (done) => {
            const fullMessage = {};
            let msg_routing_type = CMD_COMM_GROUP;

            if (internal_message) {
                msg_routing_type = CMD_TYPE_INTERMODULE;
            } else if (targetPartyID) {
                msg_routing_type = CMD_COMM_INDIVIDUAL;
            }

            fullMessage[JSON_INTERMODULE_MODULE_KEY] = this.m_module_key;
            fullMessage[ANDRUAV_PROTOCOL_TARGET_ID] = targetPartyID;
            fullMessage[INTERMODULE_ROUTING_TYPE] = msg_routing_type;
            fullMessage[ANDRUAV_PROTOCOL_MESSAGE_TYPE] = andruav_message_id;
            fullMessage[ANDRUAV_PROTOCOL_MESSAGE_CMD] = jmsg;

            const msg = JSON.stringify(fullMessage);
            console.log(`sendJMSG: ${msg}`);
            this.sendMSG(Buffer.from(msg), msg.length);
            done();
        });
    }

    sendBMSG(targetPartyID, bmsg, bmsg_length, andruav_message_id, internal_message, message_cmd) {
        this.m_lock.acquire('lock', (done) => {
            const fullMessage = {};
            let msg_routing_type = CMD_COMM_GROUP;

            if (internal_message) {
                msg_routing_type = CMD_TYPE_INTERMODULE;
            } else if (targetPartyID) {
                msg_routing_type = CMD_COMM_INDIVIDUAL;
            }

            fullMessage[JSON_INTERMODULE_MODULE_KEY] = this.m_module_key;
            fullMessage[ANDRUAV_PROTOCOL_TARGET_ID] = targetPartyID;
            fullMessage[INTERMODULE_ROUTING_TYPE] = msg_routing_type;
            fullMessage[ANDRUAV_PROTOCOL_MESSAGE_TYPE] = andruav_message_id;
            fullMessage[ANDRUAV_PROTOCOL_MESSAGE_CMD] = message_cmd;

            const json_msg = JSON.stringify(fullMessage);
            let msg = Buffer.from(json_msg + '\0');

            if (bmsg_length) {
                msg = Buffer.concat([msg, bmsg]);
            }

            this.sendMSG(msg, msg.length);
            done();
        });
    }

    sendMREMSG(command_type) {
        this.m_lock.acquire('lock', (done) => {
            const json_msg = {
                [JSON_INTERMODULE_MODULE_KEY]: this.m_module_key,
                [INTERMODULE_ROUTING_TYPE]: CMD_TYPE_INTERMODULE,
                [ANDRUAV_PROTOCOL_MESSAGE_TYPE]: "TYPE_AndruavModule_RemoteExecute",
                [ANDRUAV_PROTOCOL_MESSAGE_CMD]: { "C": command_type }
            };

            const msg = JSON.stringify(json_msg);
            this.sendMSG(Buffer.from(msg), msg.length);
            done();
        });
    }

    forwardMSG(message, datalength) {
        this.sendMSG(message, datalength);
    }

    onReceive(message, len) {
        console.log(`RX MSG: len ${len}: ${message}`);

        try {
            const jMsg = JSON.parse(message.toString());

            console.log(`RX MSG: jMsg ${JSON.stringify(jMsg)}`);

            if (!(ANDRUAV_PROTOCOL_MESSAGE_TYPE in jMsg) || !(INTERMODULE_ROUTING_TYPE in jMsg)) {
                return;
            }

            if (jMsg[INTERMODULE_ROUTING_TYPE] === CMD_TYPE_INTERMODULE) {
                if (!(ANDRUAV_PROTOCOL_MESSAGE_CMD in jMsg)) {
                    return;
                }

                const cmd = jMsg[ANDRUAV_PROTOCOL_MESSAGE_CMD];
                const messageType = jMsg[ANDRUAV_PROTOCOL_MESSAGE_TYPE];

                if (messageType === TYPE_AndruavModule_ID) {
                    const moduleID = cmd["f"];
                    if (!("f" in cmd) || !(ANDRUAV_PROTOCOL_SENDER in moduleID) || !(ANDRUAV_PROTOCOL_GROUP_ID in moduleID)) {
                        return;
                    }

                    this.m_party_id = moduleID[ANDRUAV_PROTOCOL_SENDER];
                    this.m_group_id = moduleID[ANDRUAV_PROTOCOL_GROUP_ID];

                    if (!this.m_FirstReceived) {
                        console.log(` ** Communicator Server Found: m_party_id(${this.m_party_id}) m_group_id(${this.m_group_id})`);
                        this.createJSONID(false);
                        this.m_FirstReceived = true;
                    }

                    if (this.m_OnReceive) {
                        this.m_OnReceive(message, len, jMsg);
                    }
                    return;
                } else if (messageType === TYPE_AndruavMessage_DUMMY) {
                    console.log(` TYPE_AndruavMessage_DUMMY ${message}`);
                }
            }

            if (this.m_OnReceive) {
                this.m_OnReceive(message, len, jMsg);
            }

        } catch (e) {
            console.error(`ERROR: ${e}`);
        }
    }

    appendExtraField(name, ms) {
        this.m_stdinValues[name] = ms;
    }

    createJSONID(reSend) {
        const json_msg = {};
        json_msg[INTERMODULE_ROUTING_TYPE] = CMD_TYPE_INTERMODULE;
        json_msg[ANDRUAV_PROTOCOL_MESSAGE_TYPE] = TYPE_AndruavModule_ID;

        const ms = {
            [JSON_INTERMODULE_MODULE_ID]: this.m_module_id,
            [JSON_INTERMODULE_MODULE_CLASS]: this.m_module_class,
            [JSON_INTERMODULE_MODULE_MESSAGES_LIST]: this.m_message_filter,
            [JSON_INTERMODULE_MODULE_FEATURES]: this.m_module_features,
            [JSON_INTERMODULE_MODULE_KEY]: this.m_module_key,
            [JSON_INTERMODULE_HARDWARE_ID]: this.m_hardware_serial,
            [JSON_INTERMODULE_HARDWARE_TYPE]: this.m_hardware_serial_type,
            [JSON_INTERMODULE_VERSION]: this.m_module_version,
            [JSON_INTERMODULE_RESEND]: reSend,
            [JSON_INTERMODULE_TIMESTAMP_INSTANCE]: this.m_instance_time_stamp
        };

        for (const [key, value] of Object.entries(this.m_stdinValues)) {
            ms[key] = value;
        }

        json_msg[ANDRUAV_PROTOCOL_MESSAGE_CMD] = ms;

        this.cUDPClient.setJsonId(JSON.stringify(json_msg));
    }
}

module.exports = CModule;

File: python/de_module.py


import json
import time
import threading
from enum import Enum
from messages import *
from udpClient import *

import json


MODULE_FEATURE_RECEIVING_TELEMETRY      = "R"
MODULE_FEATURE_SENDING_TELEMETRY        = "T"
MODULE_FEATURE_CAPTURE_IMAGE            = "C"
MODULE_FEATURE_CAPTURE_VIDEO            = "V"


MODULE_CLASS_COMM                       = "comm"
MODULE_CLASS_FCB                        = "fcb"
MODULE_CLASS_VIDEO                      = "camera"
MODULE_CLASS_P2P                        = "p2p"
MODULE_CLASS_GENERIC                    = "gen"

HARDWARE_TYPE_UNDEFINED = 0
HARDWARE_TYPE_CPU = 1

class CModule(object):

    _instance = None

    def __new__(cls, *args, **kwargs):
        if cls._instance is None:
            cls._instance = super(CModule, cls).__new__(cls)
        return cls._instance
    
    def __init__(self):
        self.m_module_class = ""
        self.m_module_id = ""
        self.m_module_key = ""
        self.m_module_version = ""
        self.m_message_filter = {}
        self.cUDPClient = None
        self.m_party_id = ""
        self.m_group_id = ""
        self.m_OnReceive = None
        self.m_stdinValues = {}
        self.m_FirstReceived = False
        self.m_module_features = []  # Initialize the list of module features
        self.m_hardware_serial = ""
        self.m_hardware_serial_type = ""
        self.m_instance_time_stamp = time.time()
        self.m_lock = threading.RLock()

    def init(self, target_ip, broadcasts_port, host, listening_port, chunk_size,):
        # UDP Server
        self.cUDPClient = CUDPClient()
        self.cUDPClient.init(target_ip, broadcasts_port, host, listening_port, chunk_size, self.onReceive)
        self.createJSONID(True)
        self.cUDPClient.start()
        return True

    def uninit(self):
        self.cUDPClient.stop()
        return True

    def defineModule(self, module_class, module_id, module_key, module_version, message_filter):
        self.m_module_class = module_class
        self.m_module_id = module_id
        self.m_module_key = module_key
        self.m_module_version = module_version
        self.m_message_filter = message_filter

    def add_module_features(self, feature):
        self.m_module_features.append(feature)
    
    def set_hardware(self, hardware_serial, hardware_serial_type):
        self.m_hardware_serial = hardware_serial
        self.m_hardware_serial_type = hardware_serial_type
    
    def sendMSG(self, msg, length):
        self.cUDPClient.sendMSG(msg, length)
    
    def send_sys_msg(self, jmsg, andruav_message_id):
        full_message = {
            ANDRUAV_PROTOCOL_TARGET_ID: SPECIAL_NAME_SYS_NAME,
            INTERMODULE_ROUTING_TYPE: CMD_COMM_SYSTEM,
            ANDRUAV_PROTOCOL_MESSAGE_TYPE: andruav_message_id,
            ANDRUAV_PROTOCOL_MESSAGE_CMD: jmsg
        }
        msg = json.dumps(full_message)
        self.send_msg(msg.encode(), len(msg))

    
    def sendJMSG(self, targetPartyID, jmsg, andruav_message_id, internal_message):
        """Generate and sends JSON text string

        Args:
            targetPartyID (_type_): destination unit partyid
            jmsg (_type_): _description_
            andruav_message_id (_type_): _description_
            internal_message (_type_): _description_
        """
        with self.m_lock:
            fullMessage = {}

            msg_routing_type = CMD_COMM_GROUP
            if internal_message:
                msg_routing_type = CMD_TYPE_INTERMODULE
            elif targetPartyID:
                msg_routing_type = CMD_COMM_INDIVIDUAL

            fullMessage[INTERMODULE_MODULE_KEY] = self.m_module_key
            fullMessage[ANDRUAV_PROTOCOL_TARGET_ID] = targetPartyID
            fullMessage[INTERMODULE_ROUTING_TYPE] = msg_routing_type
            fullMessage[ANDRUAV_PROTOCOL_MESSAGE_TYPE] = andruav_message_id
            fullMessage[ANDRUAV_PROTOCOL_MESSAGE_CMD] = jmsg

            msg = json.dumps(fullMessage)
            print(f"sendJMSG: {msg}")
            self.sendMSG(msg.encode(), len(msg))

    def sendBMSG(self, targetPartyID, bmsg, bmsg_length, andruav_message_id, internal_message, message_cmd):
        with self.m_lock:
            fullMessage = {}

            msg_routing_type = CMD_COMM_GROUP
            if internal_message:
                msg_routing_type = CMD_TYPE_INTERMODULE
            elif targetPartyID:
                msg_routing_type = CMD_COMM_INDIVIDUAL

            fullMessage[INTERMODULE_MODULE_KEY] = self.m_module_key
            fullMessage[ANDRUAV_PROTOCOL_TARGET_ID] = targetPartyID
            fullMessage[INTERMODULE_ROUTING_TYPE] = msg_routing_type
            fullMessage[ANDRUAV_PROTOCOL_MESSAGE_TYPE] = andruav_message_id
            fullMessage[ANDRUAV_PROTOCOL_MESSAGE_CMD] = message_cmd

            json_msg = json.dumps(fullMessage)

            msg = json_msg.encode('utf-8')
            msg += b'\0'
            if bmsg_length:
                msg += bmsg

            self.sendMSG(msg, len(msg))

    def sendMREMSG(self, command_type):
        with self.m_lock:
            json_msg = {
                INTERMODULE_MODULE_KEY: self.m_module_key,
                INTERMODULE_ROUTING_TYPE: "CMD_TYPE_INTERMODULE",
                ANDRUAV_PROTOCOL_MESSAGE_TYPE: "TYPE_AndruavModule_RemoteExecute",
                ANDRUAV_PROTOCOL_MESSAGE_CMD: {"C": command_type}
            }

            msg = json.dumps(json_msg)
            self.sendMSG(msg.encode(), len(msg))

    def forwardMSG(self, message, datalength):
        self.sendMSG(message, datalength)

    def onReceive(self, message, len):
        print(f"RX MSG: :len {len}:{message}")

        try:
            jMsg = json.loads(message)

            print(f"RX MSG: jMsg{json.dumps(jMsg)}")

            if ANDRUAV_PROTOCOL_MESSAGE_TYPE not in jMsg:
                return
            if INTERMODULE_ROUTING_TYPE not in jMsg:
                return

            if jMsg[INTERMODULE_ROUTING_TYPE] == CMD_TYPE_INTERMODULE:
                if ANDRUAV_PROTOCOL_MESSAGE_CMD not in jMsg:
                    return

                cmd = jMsg[ANDRUAV_PROTOCOL_MESSAGE_CMD]

                messageType = jMsg[ANDRUAV_PROTOCOL_MESSAGE_TYPE]
                if messageType == TYPE_AndruavModule_ID:
                    moduleID = cmd["f"]

                    if "f" not in cmd:
                        return
                    if ANDRUAV_PROTOCOL_SENDER not in moduleID:
                        return
                    if ANDRUAV_PROTOCOL_GROUP_ID not in moduleID:
                        return

                    self.m_party_id = moduleID[ANDRUAV_PROTOCOL_SENDER]
                    self.m_group_id = moduleID[ANDRUAV_PROTOCOL_GROUP_ID]

                    if not self.m_FirstReceived:
                        print(f" ** Communicator Server Found: m_party_id({self.m_party_id}) m_group_id({self.m_group_id})")
                        self.createJSONID(False)
                        self.m_FirstReceived = True

                    if self.m_OnReceive:
                        self.m_OnReceive(message, len, jMsg)

                    return

                elif messageType == TYPE_AndruavMessage_DUMMY:
                    print(f" TYPE_AndruavMessage_DUMMY {message}")

            if self.m_OnReceive:
                self.m_OnReceive(message, len, jMsg)

        except Exception as e:
            print(f"ERROR:{e}")

    def appendExtraField(self, name, ms):
        self.m_stdinValues[name] = ms

    def createJSONID(self, reSend):
        json_msg = {}
        
        json_msg[INTERMODULE_ROUTING_TYPE] = CMD_TYPE_INTERMODULE
        json_msg[ANDRUAV_PROTOCOL_MESSAGE_TYPE] = TYPE_AndruavModule_ID
        ms = {}
        
        ms[JSON_INTERMODULE_MODULE_ID] = self.m_module_id
        ms[JSON_INTERMODULE_MODULE_CLASS] = self.m_module_class
        ms[JSON_INTERMODULE_MODULE_MESSAGES_LIST] = self.m_message_filter
        ms[JSON_INTERMODULE_MODULE_FEATURES] = self.m_module_features
        ms[JSON_INTERMODULE_MODULE_KEY] = self.m_module_key
        ms[JSON_INTERMODULE_HARDWARE_ID] = self.m_hardware_serial
        ms[JSON_INTERMODULE_HARDWARE_TYPE] = self.m_hardware_serial_type
        ms[JSON_INTERMODULE_VERSION] = self.m_module_version
        ms[JSON_INTERMODULE_RESEND] = reSend
        ms[JSON_INTERMODULE_TIMESTAMP_INSTANCE] = self.m_instance_time_stamp
        
        for key, value in self.m_stdinValues.items():
            ms[key] = value
        
        json_msg[ANDRUAV_PROTOCOL_MESSAGE_CMD] = ms
        
        #print(json.dumps(json_msg, indent=4))
        
        self.cUDPClient.setJsonId(json.dumps(json_msg))
